### PR TITLE
added trait fn to test if a function is a method

### PR DESCRIPTION
**Pull request by purrie**
the trait function tests whatever a function takes its parent container as the first argument.

I'm not... uhh, glued for lack of better word to the name of the trait function but went with "isMethod" since I didn't had a better idea and it is more or less representative of what it does. 

I also intentionally do not treat anytype as valid condition to mark the function as a method when it's a first argument since usually that's not the intent when writing code.

I found this trait function tester useful in my code and found it missing in std so I thought I would offer it for inclusion.

### CODE REVIEW

**Comment by marler8997** [comment]
What about `hasMethod(comptime T: type, comptime name: []const u8) bool`?  More closely matches the `@hasDecl` builtin interface.

File: lib/std/meta/trait.zig [hunk 1]
```diff
@@ -145,6 +145,49 @@ test "isSliceOf" {
...
+            const DeclType = @TypeOf(@field(T, name));
+            switch (@typeInfo(DeclType)) {
+                .Fn => |f| {
+                    if (f.params.len == 0) return false;
+                    const Type = f.params[0].type orelse return false;
+                    switch (@typeInfo(Type)) {
+                        .Pointer => |ptr| {
+                            return ptr.child == T and ptr.size == .One;
+                        },
+                        else => return Type == T,
```
Comment (ianprime0509) [inline lib/std/meta/trait.zig:163]: Aside from just `T` and `*T`/`*const T` as the receiver type, there are other possibilities you may want to consider here: https://github.com/ziglang/zig/blob/1880c799aebed7efe524c040998b27a2b361ff43/test/behavior/fn.zig#L445-L479 Specifically, `?T`, `?*T`/`?*const T`, and `E!T` (where `E` is any error set). Additionally `[*c]T`/`[*c]const T` are supported, but not part of the linked behavior test: https://github.com/ziglang/zig/blob/1880c799aebed7efe524c040998b27a2b361ff43/src/Sema.zig#L26565-L26566

Comment (purrie) [reply lib/std/meta/trait.zig:163]: That's fair. I can add those.

**Comment by purrie** [comment]
> What about `hasMethod(comptime T: type, comptime name: []const u8) bool`? More closely matches the `@hasDecl` builtin interface.

It would lose composability with multiTrait tho.

**Comment by marler8997** [comment]
Oh I didn't know about `multiTrait`, interesting stuff.  Carry on.

**Comment by Vexu** [comment]
`meta.trait` has been removed.

**Pull request closed without merging**